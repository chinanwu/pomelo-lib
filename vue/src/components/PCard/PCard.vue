<template>
  <div
    :id="id"
    class="PCard"
    :class="{ 'PCard--flip': isFlipped }"
    tabindex="0"
    @click="handleClick"
    @keydown="handleKeyDown"
  >
    <div class="PCard__front">
      <slot name="front"></slot>
    </div>
    <div class="PCard__back">
      <slot name="back"></slot>
    </div>
  </div>
</template>

<script>
export default {
  name: "PCard",
  props: {
    id: {
      type: String,
      default: null
    },
    flipped: {
      type: Boolean,
      default: false
    }
  },
  data() {
    return {
      isFlipped: this.flipped
    };
  },
  emits: ["click"],
  methods: {
    handleClick() {
      this.isFlipped = !this.isFlipped;
      this.$emit("click");
    },
    handleKeyDown(event) {
      if (
        event &&
        event.keyCode &&
        !event.shiftKey &&
        !event.ctrlKey &&
        !event.altKey &&
        !event.metaKey &&
        (event.keyCode === 13 || event.keyCode === 32) // if enter or space is pressed
      ) {
        event.preventDefault();
        this.handleClick();
      }
    }
  }
};
</script>

<style lang="less" scoped>
.PCard {
  width: 40rem;
  height: 20rem;
  transform-origin: 50% 50%;
  transform-style: preserve-3d;
  transition: transform 0.25s ease-in-out, box-shadow 0.3s;
  cursor: pointer;

  &:focus {
    box-shadow: 0 8px 17px 2px rgba(0, 0, 0, 0.14),
      0 3px 14px 2px rgba(0, 0, 0, 0.12), 0 5px 5px -3px rgba(0, 0, 0, 0.2);
  }

  &:hover {
    box-shadow: 0 16px 24px 2px rgba(0, 0, 0, 0.14),
      0 6px 30px 5px rgba(0, 0, 0, 0.12), 0 8px 10px -7px rgba(0, 0, 0, 0.2);
  }
}

.PCard--flip {
  transform: rotateY(180deg);
}

.PCard__front,
.PCard__back {
  display: flex;
  align-items: center;
  position: absolute;
  top: 0;
  left: 0;
  backface-visibility: hidden;
  width: 100%;
  height: 100%;
  background-color: white;
}

.PCard__back {
  transform: rotateY(180deg);
}

@media only screen and (max-width: 500px) {
  .PCard {
    height: 42rem;
    width: 26rem;

    box-shadow: 0 8px 17px 2px rgba(0, 0, 0, 0.14),
      0 3px 14px 2px rgba(0, 0, 0, 0.12), 0 5px 5px -3px rgba(0, 0, 0, 0.2);
  }
}
</style>
