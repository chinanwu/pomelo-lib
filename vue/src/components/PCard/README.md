# PCard

A flippable base card component.

## Example
```html
<PCard>
    <template #front>
        I am the front of the card!
    </template>
    <template #back>
        And I, the back!
    </template>
</PCard>
```
Alternatively, you can use the hyphenated component name.
```html
<p-card></p-card>
```

## Props

| Prop | Type |  Default | Description | Validation |
| ---- | ---- | -------- | ----------- | ---------- |
| id | String | Null | Sets the id of the card container div | None |
| flipped | String | Null | Sets the default flipped state of the card | None |

