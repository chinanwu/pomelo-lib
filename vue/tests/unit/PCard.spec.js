import { mount, shallowMount } from "@vue/test-utils";

import PCard from "@/components/PCard/PCard";

describe("PCard", () => {
  describe("rendering", () => {
    it("renders card", () => {
      const wrapper = shallowMount(PCard);
      expect(wrapper.element).toMatchSnapshot();
    });

    it("renders flipped card", () => {
      const wrapper = shallowMount(PCard, {
        props: {
          id: "test",
          class: "CustomCardClass",
          flipped: true
        }
      });
      expect(wrapper.element).toMatchSnapshot();
    });

    it("renders card with front and back slots", () => {
      const wrapper = mount(PCard, {
        slots: {
          front: "<div class='test'>front</div>",
          back: "<div class='test'>front</div>"
        }
      });
      expect(wrapper.findAll(".test").length).toBe(2);
    });
  });

  describe("behaviour", () => {
    it("flips when clicked", () => {
      const wrapper = shallowMount(PCard, {
        props: {
          id: "test"
        }
      });
      expect(wrapper.element).toMatchSnapshot();
      wrapper.find("#test").trigger("click");
      expect(wrapper.emitted()).toHaveProperty("click");
    });

    it("flips when the enter or space button is pressed", () => {
      const wrapper = shallowMount(PCard, {
        props: {
          id: "test"
        }
      });
      expect(wrapper.element).toMatchSnapshot();
      wrapper.find("#test").trigger("keydown", {
        keyCode: 13
      });
      wrapper.find("#test").trigger("keydown", {
        keyCode: 32
      });
      expect(wrapper.emitted()).toHaveProperty("click");
      expect(wrapper.emitted("click").length).toBe(2);
    });
  });
});
