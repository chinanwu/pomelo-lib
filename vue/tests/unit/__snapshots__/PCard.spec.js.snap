// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PCard behaviour flips when clicked 1`] = `
<div
  class="PCard"
  id="test"
  tabindex="0"
>
  <div
    class="PCard__front"
  >
    
    
  </div>
  <div
    class="PCard__back"
  >
    
    
  </div>
</div>
`;

exports[`PCard behaviour flips when the enter or space button is pressed 1`] = `
<div
  class="PCard"
  id="test"
  tabindex="0"
>
  <div
    class="PCard__front"
  >
    
    
  </div>
  <div
    class="PCard__back"
  >
    
    
  </div>
</div>
`;

exports[`PCard rendering renders card 1`] = `
<div
  class="PCard"
  tabindex="0"
>
  <div
    class="PCard__front"
  >
    
    
  </div>
  <div
    class="PCard__back"
  >
    
    
  </div>
</div>
`;

exports[`PCard rendering renders flipped card 1`] = `
<div
  class="PCard PCard--flip CustomCardClass"
  id="test"
  tabindex="0"
>
  <div
    class="PCard__front"
  >
    
    
  </div>
  <div
    class="PCard__back"
  >
    
    
  </div>
</div>
`;
